terraform-provider-constellation/internal/provider/attestation_data_source.go

```diff
@@ -0,0 +1,301 @@
+			"maa_url": schema.StringAttribute{
+				MarkdownDescription: "For Azure only, the URL of the Microsoft Azure Attestation service",
+				Optional:            true,
+			},
+			"id": schema.StringAttribute{
+				Computed:            true,
+				MarkdownDescription: "The ID of the data source",
+			},
+			"measurements": schema.MapNestedAttribute{
+				Computed: true,
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"expected": schema.StringAttribute{
+							Computed: true,
+						},
+						"warn_only": schema.BoolAttribute{
+							Computed: true,
+						},
+					},
+				},
+			},
+			"attestation": schema.SingleNestedAttribute{
+				Computed:            true,
+				MarkdownDescription: "Only relevant for Azure SEV-SNP.",
+				Attributes: map[string]schema.Attribute{
+					"bootloader_version": schema.Int64Attribute{
+						Computed: true,
+					},
+					"tee_version": schema.Int64Attribute{
+						Computed: true,
+					},
+					"snp_version": schema.Int64Attribute{
+						Computed: true,
+					},
+					"microcode_version": schema.Int64Attribute{
+						Computed: true,
+					},
+					"azure_firmware_signer_config": schema.SingleNestedAttribute{
+						Computed: true,
+						Attributes: map[string]schema.Attribute{
+							"accepted_key_digests": schema.ListAttribute{
+								Computed:    true,
+								ElementType: types.StringType,
+							},
+							"enforcement_policy": schema.StringAttribute{
+								Computed: true,
+							},
+							"maa_url": schema.StringAttribute{
+								Computed: true,
+							},
+						},
+					},
+					"amd_root_key": schema.StringAttribute{
+						Computed: true,
+					},
+				},
+			},
+		},
+	}
+}
+
+// Read reads from the data source.
+func (d *AttestationDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data AttestationDataSourceModel
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	csp := cloudprovider.FromString(data.CSP.ValueString())
+	if csp == cloudprovider.Unknown {
+		resp.Diagnostics.AddAttributeError(
+			path.Root("csp"),
+			"Invalid CSP",
+			fmt.Sprintf("Invalid CSP: %s", data.CSP.ValueString()),
+		)
+		return
+	}
+	attestationVariant, err := variant.FromString(data.AttestationVariant.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Unknown Attestation Variant",
+			fmt.Sprintf("Unknown Attestation Variant: %s", data.AttestationVariant.ValueString()))
+		return
+	}
+	if attestationVariant.Equal(variant.AzureSEVSNP{}) || attestationVariant.Equal(variant.AWSSEVSNP{}) {
+		snpVersions, err := d.fetcher.FetchSEVSNPVersionLatest(ctx, attestationVariant)
+		if err != nil {
+			resp.Diagnostics.AddError("Fetching SNP Version numbers", err.Error())
+			return
+		}
+		tfSnpVersions := convertSNPAttestationTfStateCompatible(resp, attestationVariant, snpVersions)
```

Comment:
I don't think we should pass the resp object and modify it in the utility function here. IMO, we should rather return val, err from it and then add the error to the diagnostics in here, staying consistent with the other functions and removing side-effects from the `convertSNPAttestationTfStateCompatible` function.